cmd/gateway-main.go

```diff
@@ -31,212 +31,77 @@ import (
-      $ {{.HelpName}} mygcsprojectid
-
-`
-
-const b2GatewayTemplate = `NAME:
-  {{.HelpName}} - {{.Usage}}
-
-USAGE:
-  {{.HelpName}} {{if .VisibleFlags}}[FLAGS]{{end}}
-{{if .VisibleFlags}}
-FLAGS:
-  {{range .VisibleFlags}}{{.}}
-  {{end}}{{end}}
-ENVIRONMENT VARIABLES:
-  ACCESS:
-     MINIO_ACCESS_KEY: B2 account id.
-     MINIO_SECRET_KEY: B2 application key.
-
-  BROWSER:
-     MINIO_BROWSER: To disable web browser access, set this value to "off".
-
-EXAMPLES:
-  1. Start minio gateway server for B2 backend.
-      $ export MINIO_ACCESS_KEY=accountID
-      $ export MINIO_SECRET_KEY=applicationKey
-      $ {{.HelpName}}
-
-`
-
 var (
-	azureBackendCmd = cli.Command{
-		Name:               "azure",
-		Usage:              "Microsoft Azure Blob Storage.",
-		Action:             azureGatewayMain,
-		CustomHelpTemplate: azureGatewayTemplate,
-		Flags:              append(serverFlags, globalFlags...),
-		HideHelpCommand:    true,
-	}
-
-	s3BackendCmd = cli.Command{
-		Name:               "s3",
-		Usage:              "Amazon Simple Storage Service (S3).",
-		Action:             s3GatewayMain,
-		CustomHelpTemplate: s3GatewayTemplate,
-		Flags:              append(serverFlags, globalFlags...),
-		HideHelpCommand:    true,
-	}
-
-	gcsBackendCmd = cli.Command{
-		Name:               "gcs",
-		Usage:              "Google Cloud Storage.",
-		Action:             gcsGatewayMain,
-		CustomHelpTemplate: gcsGatewayTemplate,
-		Flags:              append(serverFlags, globalFlags...),
-		HideHelpCommand:    true,
-	}
-
-	b2BackendCmd = cli.Command{
-		Name:               "b2",
-		Usage:              "Backblaze B2.",
-		Action:             b2GatewayMain,
-		CustomHelpTemplate: b2GatewayTemplate,
-		Flags:              append(serverFlags, globalFlags...),
-		HideHelpCommand:    true,
-	}
-
 	gatewayCmd = cli.Command{
 		Name:            "gateway",
 		Usage:           "Start object storage gateway.",
 		Flags:           append(serverFlags, globalFlags...),
 		HideHelpCommand: true,
-		Subcommands:     []cli.Command{azureBackendCmd, s3BackendCmd, gcsBackendCmd, b2BackendCmd},
 	}
+
+	gatewayBackendReg = map[gatewayBackend]gatewayBackendEntry{}
 )
 
 // Represents the type of the gateway backend.
 type gatewayBackend string
 
-const (
-	azureBackend gatewayBackend = "azure"
-	s3Backend    gatewayBackend = "s3"
-	gcsBackend   gatewayBackend = "gcs"
-	b2Backend    gatewayBackend = "b2"
-	// Add more backends here.
-)
+// Represents the type of each item in gateway backend registry.
+type gatewayBackendEntry struct {
+	Name           gatewayBackend
+	Cmd            cli.Command
+	New            func(args cli.Args) (GatewayLayer, error)
+	IsExperimental bool
+}
+
+// Validate validates gateway backend entry.
+func (e *gatewayBackendEntry) Validate() error {
+	if e.Name == "" {
+		return errors.New("field 'Name' cannot be empty")
+	}
+
+	if e.New == nil {
+		return errors.New("field 'New' cannot be nil")
+	}
+
+	if e.Cmd.Action != nil {
```

Comment:
This check is sort of incorrect with the overall usage and pattern,  There is no reason we should add this check as it might be frivolous indeed. We should run a Validate() func right after the Action is registered, or just remove verifying Action.